File: server/chat/middleware.py
"""
Chat authentication middleware for the AI-Karen production chat system.
Provides JWT token validation, rate limiting, and security headers for chat endpoints.
"""

import logging
import time
import asyncio
from typing import Dict, Any, Optional, Callable, List
from datetime import datetime, timedelta
from fastapi import Request, Response, HTTPException, status, Depends
from fastapi.security import HTTPBearer, HTTPAuthorizationCredentials
from starlette.middleware.base import BaseHTTPMiddleware
from starlette.responses import JSONResponse
import jwt
import redis.asyncio as redis
from sqlalchemy.ext.asyncio import AsyncSession

from ..security import ExtensionAuthManager, get_extension_auth_manager
from ..config import settings

logger = logging.getLogger(__name__)

# Rate limiting storage (in production, use Redis)
rate_limit_store = {}
security_events = []

class ChatAuthenticationMiddleware(BaseHTTPMiddleware):
    """Middleware for chat authentication and security."""
    
    def __init__(self, app, redis_url: Optional[str] = None):
        super().__init__(app)
        self.redis_client = None
        self.auth_manager = get_extension_auth_manager()
        self.rate_limits = {
            "requests_per_minute": 60,
            "requests_per_hour": 1000,
            "message_per_minute": 30,
            "failed_auth_per_hour": 10
        }
        
        # Initialize Redis if available
        if redis_url:
            try:
                self.redis_client = redis.from_url(redis_url)
                logger.info("Redis client initialized for rate limiting")
            except Exception as e:
                logger.warning(f"Failed to initialize Redis client: {e}")
    
    async def dispatch(self, request: Request, call_next: Callable) -> Response:
        """Process request through authentication and security checks."""
        
        # Skip authentication for health checks and options
        if self._is_public_endpoint(request):
            return await call_next(request)
        
        # Add security headers
        response = await self._process_request(request, call_next)
        self._add_security_headers(response)
        
        return response
    
    def _is_public_endpoint(self, request: Request) -> bool:
        """Check if endpoint is public (no auth required)."""
        public_paths = [
            "/api/chat/health",
            "/api/docs",
            "/api/openapi.json",
            "/api/redoc"
        ]
        
        return any(request.url.path.startswith(path) for path in public_paths)
    
    def _is_development_mode(self) -> bool:
        """Check if system is in development mode."""
        import os
        return os.getenv("ENVIRONMENT", "development").lower() in ["development", "dev"]
    
    def _create_dev_user_context(self) -> Dict[str, Any]:
        """Create development user context for testing."""
        return {
            'user_id': 'dev-user',
            'tenant_id': 'dev-tenant',
            'roles': ['admin', 'user'],
            'permissions': ['chat:read', 'chat:write', 'chat:admin', 'conversation:create', 'message:send'],
            'token_type': 'development'
        }
    
    async def _process_request(self, request: Request, call_next: Callable) -> Response:
        """Process request with authentication and rate limiting."""
        
        client_ip = self._get_client_ip(request)
        user_agent = request.headers.get("user-agent", "")
        
        try:
            # Rate limiting check
            await self._check_rate_limit(request, client_ip)
            
            # Authentication check
            user_context = await self._authenticate_request(request)
            
            # Add user context to request state
            request.state.user_context = user_context
            request.state.client_ip = client_ip
            
            # Log security event
            await self._log_security_event(
                "request_authenticated",
                {
                    "user_id": user_context.get("user_id"),
                    "ip": client_ip,
                    "user_agent": user_agent,
                    "endpoint": request.url.path,
                    "method": request.method
                }
            )
            
            return await call_next(request)
            
        except HTTPException as e:
            # Log failed authentication
            await self._log_security_event(
                "authentication_failed",
                {
                    "ip": client_ip,
                    "user_agent": user_agent,
                    "endpoint": request.url.path,
                    "method": request.method,
                    "status_code": e.status_code,
                    "detail": e.detail
                }
            )
            
            # Check for brute force
            await self._check_brute_force(client_ip)
            
            return JSONResponse(
                status_code=e.status_code,
                content={"error": e.detail, "timestamp": datetime.utcnow().isoformat()}
            )
        except Exception as e:
            logger.error(f"Unexpected error in middleware: {e}")
            
            await self._log_security_event(
                "middleware_error",
                {
                    "ip": client_ip,
                    "endpoint": request.url.path,
                    "error": str(e)
                }
            )
            
            return JSONResponse(
                status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
                content={"error": "Internal server error", "timestamp": datetime.utcnow().isoformat()}
            )
    
    def _get_client_ip(self, request: Request) -> str:
        """Get client IP address from request."""
        # Check for forwarded headers
        forwarded_for = request.headers.get("x-forwarded-for")
        if forwarded_for:
            return forwarded_for.split(",")[0].strip()
        
        real_ip = request.headers.get("x-real-ip")
        if real_ip:
            return real_ip
        
        return request.client.host if request.client else "unknown"
    
    async def _check_rate_limit(self, request: Request, client_ip: str) -> None:
        """Check rate limiting for client."""
        current_time = time.time()
        endpoint = request.url.path
        
        # Different rate limits for different endpoints
        if endpoint.startswith("/api/chat/conversations") and request.method == "POST":
            # Message sending rate limit
            limit_key = f"message_rate:{client_ip}"
            limit = self.rate_limits["message_per_minute"]
            window = 60  # 1 minute
        else:
            # General API rate limit
            limit_key = f"api_rate:{client_ip}"
            limit = self.rate_limits["requests_per_minute"]
            window = 60  # 1 minute
        
        # Check rate limit using Redis or in-memory store
        if self.redis_client:
            await self._check_redis_rate_limit(limit_key, limit, window)
        else:
            self._check_memory_rate_limit(limit_key, limit, window, current_time)
    
    async def _check_redis_rate_limit(self, key: str, limit: int, window: int) -> None:
        """Check rate limit using Redis."""
        try:
            if self.redis_client is None:
                raise Exception("Redis client not initialized")
                
            current = await self.redis_client.incr(key)
            
            if current == 1:
                # Set expiration on first request
                await self.redis_client.expire(key, window)
            
            if current > limit:
                raise HTTPException(
                    status_code=status.HTTP_429_TOO_MANY_REQUESTS,
                    detail=f"Rate limit exceeded. Maximum {limit} requests per {window} seconds."
                )
        except Exception as e:
            logger.warning(f"Redis rate limit check failed: {e}")
            # Fallback to memory-based rate limiting
    
    def _check_memory_rate_limit(self, key: str, limit: int, window: int, current_time: float) -> None:
        """Check rate limit using in-memory store."""
        if key not in rate_limit_store:
            rate_limit_store[key] = []
        
        # Remove old requests outside the window
        rate_limit_store[key] = [
            req_time for req_time in rate_limit_store[key]
            if current_time - req_time < window
        ]
        
        # Check if limit exceeded
        if len(rate_limit_store[key]) >= limit:
            raise HTTPException(
                status_code=status.HTTP_429_TOO_MANY_REQUESTS,
                detail=f"Rate limit exceeded. Maximum {limit} requests per {window} seconds."
            )
        
        # Add current request
        rate_limit_store[key].append(current_time)
    
    async def _authenticate_request(self, request: Request) -> Dict[str, Any]:
        """Authenticate request using JWT token or API key."""
        
        # DEVELOPMENT MODE: Check if in development and provide dev user context
        if self._is_development_mode():
            logger.debug("Development mode detected: Using development user context")
            return self._create_dev_user_context()
        
        # Try JWT token authentication first
        authorization = request.headers.get("authorization")
        if authorization and authorization.startswith("Bearer "):
            token = authorization.split(" ")[1]
            return await self._validate_jwt_token(token)
        
        # Try API key authentication
        api_key = request.headers.get("x-api-key")
        if api_key:
            return await self._authenticate_api_key(api_key)
        
        # Try cookie-based authentication
        token = request.cookies.get("access_token")
        if token:
            return await self._validate_jwt_token(token)
        
        # No authentication provided
        raise HTTPException(
            status_code=status.HTTP_401_UNAUTHORIZED,
            detail="Authentication required"
        )
    
    async def _validate_jwt_token(self, token: str) -> Dict[str, Any]:
        """Validate JWT token and return user context."""
        try:
            # Use existing auth manager to validate token
            from fastapi import HTTPException
            from fastapi.security import HTTPAuthorizationCredentials
            from fastapi import Request as FastAPIRequest
            
            # Create mock request for the auth manager
            mock_request = FastAPIRequest(
                scope={
                    "type": "http",
                    "method": "GET",
                    "path": "/api/chat",
                    "headers": {}
                }
            )
            
            # Create mock credentials for the auth manager
            credentials = HTTPAuthorizationCredentials(scheme="Bearer", credentials=token)
            
            # Authenticate using existing extension auth manager
            user_context = await self.auth_manager.authenticate_extension_request(
                request=mock_request,
                credentials=credentials
            )
            
            # Add chat-specific permissions
            await self._add_chat_permissions(user_context)
            
            return user_context
            
        except jwt.ExpiredSignatureError:
            raise HTTPException(
                status_code=status.HTTP_401_UNAUTHORIZED,
                detail="Token expired"
            )
        except jwt.InvalidTokenError as e:
            raise HTTPException(
                status_code=status.HTTP_401_UNAUTHORIZED,
                detail=f"Invalid token: {str(e)}"
            )
        except Exception as e:
            logger.error(f"Token validation error: {e}")
            raise HTTPException(
                status_code=status.HTTP_401_UNAUTHORIZED,
                detail="Authentication failed"
            )
    
    async def _authenticate_api_key(self, api_key: str) -> Dict[str, Any]:
        """Authenticate using API key."""
        try:
            # Use existing auth manager for API key authentication
            user_context = await self.auth_manager._authenticate_with_api_key(api_key)
            
            # Add chat-specific permissions
            await self._add_chat_permissions(user_context)
            
            return user_context
            
        except Exception as e:
            raise HTTPException(
                status_code=status.HTTP_401_UNAUTHORIZED,
                detail="Invalid API key"
            )
    
    async def _add_chat_permissions(self, user_context: Dict[str, Any]) -> None:
        """Add chat-specific permissions to user context."""
        existing_permissions = user_context.get("permissions", [])
        
        # Add base chat permissions
        chat_permissions = [
            "chat:read",
            "chat:write",
            "chat:conversations:read",
            "chat:conversations:write",
            "chat:messages:read",
            "chat:messages:write"
        ]
        
        # Add admin permissions if user is admin
        if "admin" in user_context.get("roles", []):
            chat_permissions.extend([
                "chat:admin",
                "chat:providers:read",
                "chat:providers:write",
                "chat:users:read",
                "chat:audit:read"
            ])
        
        # Merge permissions without duplicates
        all_permissions = list(set(existing_permissions + chat_permissions))
        user_context["permissions"] = all_permissions
    
    async def _check_brute_force(self, client_ip: str) -> None:
        """Check for brute force attacks and block if necessary."""
        current_time = time.time()
        window = 3600  # 1 hour
        max_failed = self.rate_limits["failed_auth_per_hour"]
        
        # Track failed attempts
        key = f"failed_auth:{client_ip}"
        
        if self.redis_client:
            failed_count = await self.redis_client.incr(key)
            
            if failed_count == 1:
                await self.redis_client.expire(key, window)
            
            if failed_count > max_failed:
                await self._log_security_event(
                    "brute_force_detected",
                    {
                        "ip": client_ip,
                        "failed_count": failed_count,
                        "action": "temporary_block"
                    }
                )
                
                raise HTTPException(
                    status_code=status.HTTP_429_TOO_MANY_REQUESTS,
                    detail="Too many failed authentication attempts. Please try again later."
                )
        else:
            # In-memory fallback
            if key not in rate_limit_store:
                rate_limit_store[key] = []
            
            # Clean old attempts
            rate_limit_store[key] = [
                req_time for req_time in rate_limit_store[key]
                if current_time - req_time < window
            ]
            
            if len(rate_limit_store[key]) >= max_failed:
                await self._log_security_event(
                    "brute_force_detected",
                    {
                        "ip": client_ip,
                        "failed_count": len(rate_limit_store[key]),
                        "action": "temporary_block"
                    }
                )
                
                raise HTTPException(
                    status_code=status.HTTP_429_TOO_MANY_REQUESTS,
                    detail="Too many failed authentication attempts. Please try again later."
                )
            
            rate_limit_store[key].append(current_time)
    
    async def _log_security_event(self, event_type: str, data: Dict[str, Any]) -> None:
        """Log security events for monitoring."""
        event = {
            "timestamp": datetime.utcnow().isoformat(),
            "event_type": event_type,
            "data": data
        }
        
        # Add to in-memory store (in production, use proper logging)
        security_events.append(event)
        
        # Keep only last 1000 events in memory
        if len(security_events) > 1000:
            security_events.pop(0)
        
        # Log to standard logger
        logger.info(f"Security event: {event_type} - {data}")
    
    def _add_security_headers(self, response: Response) -> None:
        """Add security headers to response."""
        # Security headers
        response.headers["X-Content-Type-Options"] = "nosniff"
        response.headers["X-Frame-Options"] = "DENY"
        response.headers["X-XSS-Protection"] = "1; mode=block"
        response.headers["Strict-Transport-Security"] = "max-age=31536000; includeSubDomains"
        response.headers["Referrer-Policy"] = "strict-origin-when-cross-origin"
        response.headers["Content-Security-Policy"] = (
            "default-src 'self'; "
            "script-src 'self' 'unsafe-inline' 'unsafe-eval'; "
            "style-src 'self' 'unsafe-inline'; "
            "img-src 'self' data: https:; "
            "font-src 'self'; "
            "connect-src 'self' wss:; "
            "frame-ancestors 'none';"
        )


# Dependency functions for route protection
async def get_current_chat_user(request: Request) -> Dict[str, Any]:
    """Get current authenticated user from request state."""
    if not hasattr(request.state, "user_context"):
        raise HTTPException(
            status_code=status.HTTP_401_UNAUTHORIZED,
            detail="Authentication required"
        )
    
    return request.state.user_context


async def require_chat_permission(permission: str):
    """Dependency to require specific chat permission."""
    async def permission_checker(
        request: Request,
        user_context: Dict[str, Any] = Depends(get_current_chat_user)
    ) -> Dict[str, Any]:
        user_permissions = user_context.get("permissions", [])
        
        if permission not in user_permissions and "chat:admin" not in user_permissions:
            raise HTTPException(
                status_code=status.HTTP_403_FORBIDDEN,
                detail=f"Insufficient permissions. Required: {permission}"
            )
        
        return user_context
    
    return permission_checker


async def require_chat_admin(
    user_context: Dict[str, Any] = Depends(get_current_chat_user)
) -> Dict[str, Any]:
    """Require chat admin permissions."""
    user_permissions = user_context.get("permissions", [])
    
    if "chat:admin" not in user_permissions and "admin" not in user_context.get("roles", []):
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail="Chat admin permissions required"
        )
    
    return user_context


# Rate limiting utilities
async def check_message_rate_limit(user_id: str, redis_client=None) -> bool:
    """Check if user has exceeded message rate limit."""
    limit = 30  # messages per minute
    
    if redis_client:
        key = f"message_rate:{user_id}"
        current = await redis_client.incr(key)
        
        if current == 1:
            await redis_client.expire(key, 60)
        
        return current <= limit
    else:
        # In-memory fallback
        current_time = time.time()
        key = f"message_rate:{user_id}"
        
        if key not in rate_limit_store:
            rate_limit_store[key] = []
        
        # Clean old messages
        rate_limit_store[key] = [
            msg_time for msg_time in rate_limit_store[key]
            if current_time - msg_time < 60
        ]
        
        if len(rate_limit_store[key]) >= limit:
            return False
        
        rate_limit_store[key].append(current_time)
        return True


# Security event retrieval
async def get_security_events(limit: int = 100) -> List[Dict[str, Any]]:
    """Get recent security events."""
    return security_events[-limit:]


# CORS configuration
def get_chat_cors_config() -> Dict[str, Any]:
    """Get CORS configuration for chat endpoints."""
    return {
        "allow_origins": settings.get("chat.cors_origins", ["http://localhost:3000"]),
        "allow_credentials": True,
        "allow_methods": ["GET", "POST", "PUT", "DELETE", "OPTIONS"],
        "allow_headers": [
            "Authorization",
            "Content-Type",
            "X-API-Key",
            "X-Requested-With"
        ]
    }

File: ui_launchers/KAREN-Theme-Enterprise/src/middleware.ts
import { NextResponse } from 'next/server';
import type { NextRequest } from 'next/server';

// Rate limiting store (in-memory for development - use Redis in production)
const rateLimit = new Map<string, { count: number; resetTime: number }>();
const RATE_LIMIT_MAX = 100; // Max requests per window
const RATE_LIMIT_WINDOW = 60000; // 1 minute in ms

// Get allowed origins from environment
const getAllowedOrigins = (): string[] => {
  const allowedOriginsEnv = process.env.ALLOWED_ORIGINS || '';
  if (allowedOriginsEnv) {
    return allowedOriginsEnv.split(',').map(origin => origin.trim());
  }
  
  // Default: localhost for development
  const isDev = process.env.NODE_ENV !== 'production';
  return isDev 
    ? ['http://localhost:3000', 'http://localhost:9002', 'http://localhost:3001']
    : []; // Production must explicitly set ALLOWED_ORIGINS
};

// Check if origin is allowed
const isOriginAllowed = (origin: string | null): boolean => {
  if (!origin) return false;
  const allowedOrigins = getAllowedOrigins();
  
  // In production, require explicit ALLOWED_ORIGINS configuration
  if (process.env.NODE_ENV === 'production' && allowedOrigins.length === 0) {
    console.error('SECURITY: ALLOWED_ORIGINS environment variable must be set in production');
    return false;
  }
  
  return allowedOrigins.includes(origin);
};

// Rate limiting check
const checkRateLimit = (identifier: string): boolean => {
  const now = Date.now();
  const record = rateLimit.get(identifier);
  
  if (!record || now > record.resetTime) {
    // Create new rate limit window
    rateLimit.set(identifier, {
      count: 1,
      resetTime: now + RATE_LIMIT_WINDOW
    });
    return true;
  }
  
  if (record.count >= RATE_LIMIT_MAX) {
    return false; // Rate limit exceeded
  }
  
  record.count++;
  return true;
};

// Get client identifier for rate limiting
const getClientIdentifier = (request: NextRequest): string => {
  // Use forwarded IP if available (behind reverse proxy)
  const forwarded = request.headers.get('x-forwarded-for');
  const ip = forwarded ? forwarded.split(',')[0].trim() : request.ip;
  return ip || 'unknown';
};

export function middleware(request: NextRequest) {
  const { pathname } = request.nextUrl;
  const response = NextResponse.next();

  // ============================================================
  // SECURITY: Rate limiting for all routes
  // ============================================================
  const clientId = getClientIdentifier(request);
  if (!checkRateLimit(clientId)) {
    return new NextResponse(
      JSON.stringify({
        error: 'Too many requests',
        message: 'Rate limit exceeded. Please try again later.'
      }),
      {
        status: 429,
        headers: {
          'Content-Type': 'application/json',
          'Retry-After': '60'
        }
      }
    );
  }

  // Handle chunk loading errors by redirecting to a safe route
  if (pathname.includes('_next/static/chunks') && request.method === 'GET') {
    // Log the missing chunk for debugging
    // If it's a page chunk that's missing, we can try to regenerate it
    if (pathname.includes('/app/') && pathname.endsWith('.js')) {
      // Return a 404 response that Next.js can handle gracefully
      return new NextResponse(null, { status: 404 });
    }
  }

  // ============================================================
  // SECURITY: CORS restrictions for API routes
  // ============================================================
  if (pathname.startsWith('/api/')) {
    const origin = request.headers.get('origin');
    
    // Handle preflight OPTIONS requests
    if (request.method === 'OPTIONS') {
      if (isOriginAllowed(origin)) {
        return new NextResponse(null, {
          status: 200,
          headers: {
            'Access-Control-Allow-Origin': origin || '',
            'Access-Control-Allow-Methods': 'GET, POST, PUT, DELETE, OPTIONS',
            'Access-Control-Allow-Headers': 'Content-Type, Authorization',
            'Access-Control-Allow-Credentials': 'true',
            'Access-Control-Max-Age': '86400' // 24 hours
          }
        });
      }
      return new NextResponse(null, { status: 403 }); // Forbidden origin
    }
    
    // Apply CORS for regular API requests
    if (isOriginAllowed(origin)) {
      response.headers.set('Access-Control-Allow-Origin', origin || '');
      response.headers.set('Access-Control-Allow-Credentials', 'true');
      response.headers.set('Access-Control-Expose-Headers', 'Content-Length, Content-Type');
    } else {
      // Origin not allowed
      return new NextResponse(
        JSON.stringify({
          error: 'Forbidden',
          message: 'Origin not allowed'
        }),
        {
          status: 403,
          headers: { 'Content-Type': 'application/json' }
        }
      );
    }
    
    response.headers.set('Access-Control-Allow-Methods', 'GET, POST, PUT, DELETE, OPTIONS');
    response.headers.set('Access-Control-Allow-Headers', 'Content-Type, Authorization');
    // Don't cache API responses by default
    response.headers.set('Cache-Control', 'no-store, must-revalidate');
    return response;
  }

  // Add performance and back/forward cache support headers
  // ============================================================
  // These headers help the browser cache pages for instant back/forward navigation
  // SECURITY: Add Content Security Policy
  // ============================================================
  if (!pathname.startsWith('/api/')) {
    // Content Security Policy
    const cspDirectives = [
      "default-src 'self'",
      "script-src 'self' 'unsafe-inline' 'unsafe-eval'", // unsafe-inline needed for Next.js dev mode
      "style-src 'self' 'unsafe-inline'", // unsafe-inline needed for styled-jsx
      "img-src 'self' data: https: blob:",
      "font-src 'self' data:",
      "connect-src 'self' https://api.yourdomain.com", // Replace with actual backend URL
      "frame-ancestors 'none'",
      "base-uri 'self'",
      "form-action 'self'"
    ].join('; ');
    
    response.headers.set('Content-Security-Policy', cspDirectives);
    response.headers.set('X-Frame-Options', 'DENY');
    response.headers.set('X-Content-Type-Options', 'nosniff');
    response.headers.set('Referrer-Policy', 'strict-origin-when-cross-origin');
    
    // Enable back/forward cache (bfcache) support
    response.headers.set('Cache-Control', 'public, max-age=0, must-revalidate');

    // Prevent unload handlers that block bfcache
    // Note: Your app should avoid using beforeunload/unload events

    // Add timing headers for performance monitoring
    response.headers.set('Server-Timing', 'middleware;dur=0');
  }

  // ============================================================
  // AUTHENTICATION: Protected routes
  // ============================================================
  // TODO: Implement authentication check for protected routes
  // Example: if (pathname.startsWith('/admin') && !isAuthenticated(request)) return redirectToLogin(request);
  return response;
}

export const config = {
  matcher: [
    /*
     * Match all request paths except for the ones starting with:
     * - _next/static (static files)
     * - _next/image (image optimization files)
     * - favicon.ico (favicon file)
     */
    '/((?!_next/static|_next/image|favicon.ico).*)',
  ],
};
